File: TowerDefence.spritebuilder/Engine/PlaceEngine.swift
//
//  PlaceEngine.swift
//  TowerDefence
//
//  Created by koveres on 08.12.14.
//  Copyright (c) 2014 Apportable. All rights reserved.
//

import Foundation



public class PlaceEngine {
    
    
    // List<Double> placesToSetTower;
    var areaToSetTower:Area;
    var areaSetEdTower:Area;
    
    
    
    public init( placesToSetTower:[CGPoint]) {
        areaToSetTower =    Area();
        areaSetEdTower =    Area();
        calcArea(placesToSetTower);
    }
    
    
    func canAddTower( newTower:Tower) -> (Bool) {
    if(!testOnEmptyPlace(newTower))
    {return  false;}
    if(!testOnNearTower(newTower))
    {return false;}
    return true;
    }
    
       func testOnEmptyPlace( newTower:Tower) -> (Bool) {
    
        var rect:CGRect = newTower.getNeedPlace();
        return areaToSetTower.contains(rect);
    
    }
    
    private func calcArea(  placesToSetTower:[CGPoint]) {
        var sizePlacePoint:CDouble = StaticSettings.sizePlacePoint;
    
    
    for    double1  in placesToSetTower  {
        var xx:Double = Double(Double(double1.x) - Double(sizePlacePoint/2.0));
        var xy:Double = Double(Double(double1.y) - Double(sizePlacePoint/2.0));
        
        var rect:CGRect = CGRect(x: xx, y: xy, width: Double(sizePlacePoint), height: Double(sizePlacePoint));
    //    CGRect( Double(double1.x - sizePlacePoint/2),  Double(double1.y - sizePlacePoint/2), 
        //Double(sizePlacePoint), Double(sizePlacePoint));
        var newarea:Area =    Area(p: rect);
    areaToSetTower.add(newarea);
    }
    }
    
    func addTower( newTower:Tower) { 
        var newarea:Area =    Area(
            p: newTower.getNeedPlace());
        areaSetEdTower.add(newarea);
    }
    
    private func testOnNearTower( newTower: Tower) -> (Bool) {
        var rect:CGRect = newTower.getNeedPlace();
        return !areaSetEdTower.intersects(rect);
    }
    
    
    
    
    
    
    
}

File: test.spritebuilder/Source/TDEngine/Engine/PlaceEngine.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package Engine;

import Unit.Tower.Tower;
import Utils.Point2D;
import internalObject.StaticSettings;
import Utils.MyGeom.Polygon;
import Utils.MyGeom.Shape;
import Utils.MyGeom.Area;
 
 
import Utils.MyGeom.Rectangle2D;
import java.util.List;

/**
 *
 * @author vereshchagin
 */
    class PlaceEngine {

    
   // List<Double> placesToSetTower;
    Utils.MyGeom.Area areaToSetTower;
    Utils.MyGeom.Area areaSetEdTower;
   
    public PlaceEngine(List<List<Point2D>> placesToSetTower,byte notneedParamForUndestendThatThisIsThePOligonCOnstructor) {
        areaToSetTower = new  Area();
        areaSetEdTower = new  Area();
        for (List<Point2D> list : placesToSetTower) {
           addPoligon(list);  
        }
    }
    
    public PlaceEngine(List<Point2D> placesToSetTower) {
        areaToSetTower = new  Area();
        areaSetEdTower = new  Area();
        calcArea(placesToSetTower);
    }
  

    boolean canAddTower(Tower newTower) {
        if(!testOnEmptyPlace(newTower))
            return  false;
        if(!testOnNearTower(newTower))
            return false;
        return true;
    }

    private boolean testOnEmptyPlace(Tower newTower) {
        
        Utils.MyGeom.Rectangle2D rect = newTower.getNeedPlace();
        
        return areaToSetTower.contains(rect);
         
    }
    private void addPoligon(List<Point2D> placesToSetTower) {
        int size = placesToSetTower.size();
       int[] arrx = new int[size]; 
       int[] arry = new int[size]; 
 
              int i = 0; 
        for (Point2D point2D : placesToSetTower) {
            arrx[i] = (int) point2D.x;
            arry[i] = (int) point2D.y; 
            i++;
        }
        Shape p;
            p = new Polygon(arrx,arry,size);
         areaToSetTower.add(new Area(p));   
    }
    private void calcArea(List<Point2D> placesToSetTower) {
        double sizePlacePoint = StaticSettings.sizePlacePoint;
       
       
        for (Point2D double1 : placesToSetTower) {
            Area newarea = new  Area(new Rectangle2D(
                     double1.x - sizePlacePoint/2,double1.y - sizePlacePoint/2,
                      sizePlacePoint, sizePlacePoint
                     ));
            areaToSetTower.add(newarea); 
        }
    }

    void addTower(Tower newTower) {
       
        Area newarea = new  Area( 
                 newTower.getNeedPlace()
                    );
      areaSetEdTower.add(newarea);
    }

    private boolean testOnNearTower(Tower newTower) {
        Rectangle2D rect = newTower.getNeedPlace();       
        return !areaSetEdTower.intersects(rect);
    }

    
    
    

     
    
}
